File: android/src/main/java/com/mahakarshan/telnetconnection/TelnetConnectionPlugin.java
package com.example.telnetconnection;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.PluginMethod;

import org.apache.commons.net.telnet.TelnetClient;

import java.io.InputStream;
import java.io.OutputStream;
import java.io.PrintStream;
import java.util.Scanner;

@CapacitorPlugin(name = "TelnetConnection")
public class TelnetConnectionPlugin extends Plugin {

    private TelnetClient telnetClient;
    private InputStream inputStream;
    private PrintStream outputStream;
    private Thread readThread;

    @PluginMethod
    public void connect(PluginCall call) {
        String host = call.getString("host");
        int port = call.getInt("port", 23); // Default Telnet port is 23

        try {
            telnetClient = new TelnetClient();
            telnetClient.connect(host, port);

            inputStream = telnetClient.getInputStream();
            outputStream = new PrintStream(telnetClient.getOutputStream());

            // Start a background thread to listen to data from the server
            readThread = new Thread(() -> {
                try (Scanner scanner = new Scanner(inputStream)) {
                    while (telnetClient.isConnected() && scanner.hasNextLine()) {
                        String serverData = scanner.nextLine();
                        JSObject ret = new JSObject();
                        ret.put("data", serverData);
                        notifyListeners("telnetDataReceived", ret); // Emit event for data received
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
            readThread.start();

            call.resolve();
        } catch (Exception e) {
            call.reject("Connection failed", e);
        }
    }

    @PluginMethod
    public void sendCommand(PluginCall call) {
        String command = call.getString("command");

        if (telnetClient != null && telnetClient.isConnected()) {
            outputStream.println(command);
            outputStream.flush();
            call.resolve();
        } else {
            call.reject("Not connected");
        }
    }

    @PluginMethod
    public void disconnect(PluginCall call) {
        try {
            if (telnetClient != null && telnetClient.isConnected()) {
                telnetClient.disconnect();
                if (readThread != null && readThread.isAlive()) {
                    readThread.interrupt();
                }
                call.resolve();
            }
        } catch (Exception e) {
            call.reject("Failed to disconnect", e);
        }
    }
}

File: ios/Sources/telentConnectionPlugin/TelnetConnectionPlugin.swift
import Foundation
import Capacitor

@objc(TelnetConnectionPlugin)
public class TelnetConnectionPlugin: CAPPlugin {
    
    private let implementation = TelnetClient() // Telnet Client implementation

    // Method to establish a Telnet connection
    @objc func connect(_ call: CAPPluginCall) {
        guard let host = call.getString("host") else {
            call.reject("Host is required")
            return
        }
        let port = call.getInt("port") ?? 23 // Default Telnet port is 23

        DispatchQueue.global(qos: .background).async {
            do {
                try self.implementation.connect(host: host, port: port)
                call.resolve() // Successfully connected
            } catch {
                call.reject("Connection failed")
            }
        }
    }

    // Method to send commands to the Telnet server (e.g., username, password)
    @objc func sendCommand(_ call: CAPPluginCall) {
        guard let command = call.getString("command") else {
            call.reject("Command is required")
            return
        }
        
        // Send the command
        if self.implementation.isConnected {
            self.implementation.sendCommand(command: command)
            call.resolve() // Successfully sent command
        } else {
            call.reject("Not connected to the Telnet server")
        }
    }

    // Method to listen for Telnet data
    @objc func listenForData(_ call: CAPPluginCall) {
        if self.implementation.isConnected {
            DispatchQueue.global(qos: .background).async {
                let data = self.implementation.listenForData()
                call.resolve([
                    "data": data
                ])
            }
        } else {
            call.reject("Not connected to the Telnet server")
        }
    }
}
